walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,108 +14,255 @@
-                           .filter(e -> request.getAttendees()
-                                               .containsAll(e.getAttendees()))
+                           .filter(event -> !Collections.disjoint(request.getAttendees(), event.getAttendees()))
+                           .collect(Collectors.toList());
+    List<Event> optionalEvents = eventsList.stream()
+                           .filter(event -> !Collections.disjoint(request.getOptionalAttendees(), event.getAttendees()))
+
                            .collect(Collectors.toList());
     
-    Collection<TimeRange> allTRs = queryHelper(eventsList, request);
+    Collection<TimeRange> allTimeRanges = findOpenTimeRanges(eventsList, request);
 
-    // If there are no attendes, there could still be optional attendees
-    // so all valid {@code TimeRange} objects are returned.
-    if (request.getAttendees().isEmpty()) {
-      return allTRs;
-    }
+    Collection<TimeRange> mandatoryTRs = findOpenTimeRanges(mandatoryEvents, request);
 
-    Collection<TimeRange> mandatoryTRs = queryHelper(mandatoryEvents, request);
+    Collection<TimeRange> maxOptions = findPossibleTimeSlots(optionalEvents,
+                                                             mandatoryTRs, 
+                                                             request);
 
-    // If there are no avilable {@code TimeRange} objects for all attendees,
-    // avilable TRs for mandatory attendees should be returned in case there
-    // are avilable TRs for them.
-    if (allTRs.isEmpty()) {
-        return mandatoryTRs;
+    if (allTimeRanges.isEmpty() && !mandatoryTRs.isEmpty()) {
+      if (maxOptions.isEmpty()) {
+          return mandatoryTRs;
+      }
+      return maxOptions;
     }
 
-    return allTRs;
+    if (request.getAttendees().isEmpty()) {
+      return allTimeRanges;
+    }
+    
+    if (mandatoryTRs.isEmpty() && !request.getOptionalAttendees().isEmpty()) {
+      return filterMaxOptions(Arrays.asList(TimeRange.WHOLE_DAY), 
+                        optionalEvents, 
+                        request);
 
+    }
 
+    return allTimeRanges;
   }
 
   /** 
    * This is a helper method for .query(). 
    * It works by finding all the availabe {@code TimeRange} objects that 
    * qualify for request.
    */
-  public Collection<TimeRange> queryHelper(Collection<Event> separateEvents, MeetingRequest request) {
+  public Collection<TimeRange> findOpenTimeRanges(Collection<Event> separateEvents, MeetingRequest request) {
     // Map each {@code Event} to it's {@code TimeRange}.
-    List<TimeRange> TimeRangeList = separateEvents.stream()
-                     .map(a -> a.getWhen())
+    List<TimeRange> earliestStartTimes = separateEvents.stream()
+                     .map(Event::getWhen)
                      .collect(Collectors.toList());
 
     // Resulting Collection
     Collection<TimeRange> openTimeRanges = new ArrayList();
 
     int currentMinTime = 0;
```

Comment:
Start with `TimeRange.START_OF_DAY` instead of `0`

Reply:
Done.